rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNode.java

```diff
@@ -0,0 +1,162 @@
+            ctx.enqueueForTellNext(tbMsg, relationType);
+        }
+    }
+
+    private ListenableFuture<List<EntityId>> findMatchedZones(TbContext tbContext, TbMsg tbMsg) throws TbNodeException {
+        JsonObject msgDataObj = getJsonObject(tbMsg);
+        double latitude = getValueFromMessageByName(tbMsg, msgDataObj, config.getLatitudeKeyName());
+        double longitude = getValueFromMessageByName(tbMsg, msgDataObj, config.getLongitudeKeyName());
+
+        if (config.getRelationsQuery() == null) {
+            return Futures.immediateFuture(Collections.emptyList());
+        }
+
+        ListenableFuture<List<EntityId>> entityIdsFuture = EntitiesRelatedEntityIdAsyncLoader.findEntitiesAsync(tbContext, tbMsg.getOriginator(), config.getRelationsQuery());
+
+        ListenableFuture<List<Pair<EntityId, Optional<AttributeKvEntry>>>> entityIdToAttributesFuture = Futures.transformAsync(entityIdsFuture, relatedZones -> {
+            List<ListenableFuture<Pair<EntityId, Optional<AttributeKvEntry>>>> attributeFutures = relatedZones.stream()
+                    .map(relatedGeofenceId -> Futures.transform(
+                            tbContext.getAttributesService()
+                                    .find(tbContext.getTenantId(), relatedGeofenceId, AttributeScope.SERVER_SCOPE, config.getPerimeterAttributeKey()),
+                            attribute -> Pair.of(relatedGeofenceId, attribute),
+                            MoreExecutors.directExecutor()
+                    ))
+                    .collect(Collectors.toList());
+
+            return Futures.allAsList(attributeFutures);
+        }, MoreExecutors.directExecutor());
+
+        return Futures.transformAsync(entityIdToAttributesFuture, entityIdAttributePairs -> {
+            List<EntityId> matchedGeofenceIds = new ArrayList<>();
+
+            entityIdAttributePairs.forEach(pair -> {
+                EntityId entityId = pair.getLeft();
+                Optional<AttributeKvEntry> attributeKvEntry = pair.getRight();
+
+                attributeKvEntry.map(this::mapToPerimeter).ifPresent(perimeter -> {
+                    try {
+                        if (checkMatches(perimeter, latitude, longitude)) {
+                            matchedGeofenceIds.add(entityId);
+                        }
+                    } catch (TbNodeException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            });
+
+            return Futures.immediateFuture(matchedGeofenceIds);
+        }, MoreExecutors.directExecutor());
+    }
+
+
+    private Perimeter mapToPerimeter(AttributeKvEntry attributeKvEntry) {
+        Optional<String> jsonValue = attributeKvEntry.getJsonValue();
+        if (jsonValue.isEmpty()) {
+            return null;
+        }
+        JsonNode jsonNode = JacksonUtil.toJsonNode(jsonValue.get());
+        return JacksonUtil.treeToValue(jsonNode, Perimeter.class);
```

Comment:
please check if this can be replaced with JacksonUtil.fromString() method

Reply:
I used fromString, check one more time